File: challenges/2026-03-03-json-config-validator.py
"""
Daily Challenge 2026-03-03 — JSON Config Validator
Language: Python | Difficulty: Beginner

=== Background ===
Configuration files are everywhere — web apps, CLI tools, game settings.
Most use JSON because it's readable and widely supported. But loading a JSON
file doesn't mean the *contents* are valid — missing keys, wrong types, and
bad values can cause confusing crashes deep in your code.

Today you'll write a validator that checks a config dict against a schema,
and produces clear, human-friendly error messages.

=== Your Task ===
Implement `validate_config(config: dict, schema: dict) -> list[str]`.

The schema describes what keys are required and what type each value must be:

    schema = {
        "host":    {"type": str,  "required": True},
        "port":    {"type": int,  "required": True},
        "debug":   {"type": bool, "required": False},
        "timeout": {"type": float, "required": False},
    }

Rules:
1. Every key marked `"required": True` must be present in config.
2. If a key is present, its value must be an instance of the given type.
   (Note: in Python, `bool` is a subclass of `int`. If the schema says `int`
    but the value is `True` or `False`, that should be treated as a TYPE ERROR
    — a bool is not a valid int for our purposes.)
3. Collect ALL errors (don't stop at the first one).
4. Return a list of error strings. Return an empty list if config is valid.

Error message format (use these exactly):
  - Missing required key:   "Missing required key: 'host'"
  - Wrong type:             "Key 'port': expected int, got str"

=== Hints ===
- Use `isinstance(value, expected_type)` for type checking.
- Remember the special bool/int case — check for `bool` first!
- Iterate over the schema, not the config, so you catch missing keys.

=== Examples ===

Example 1 — valid config:
    config = {"host": "localhost", "port": 8080, "debug": True}
    schema = {
        "host":  {"type": str,  "required": True},
        "port":  {"type": int,  "required": True},
        "debug": {"type": bool, "required": False},
    }
    validate_config(config, schema)  # → []

Example 2 — multiple errors:
    config = {"host": 1234, "debug": "yes"}
    schema = {
        "host":  {"type": str,  "required": True},
        "port":  {"type": int,  "required": True},
        "debug": {"type": bool, "required": False},
    }
    validate_config(config, schema)
    # → ["Key 'host': expected str, got int",
    #    "Missing required key: 'port'",
    #    "Key 'debug': expected bool, got str"]

=== Starter Code ===
"""

def validate_config(config: dict, schema: dict) -> list[str]:
    """
    Validate config against schema.

    Args:
        config: The configuration dict to validate.
        schema: A dict mapping key names to {"type": <type>, "required": <bool>}.

    Returns:
        A list of error strings (empty list = valid).
    """
    errors = []

    # TODO: Iterate over the schema entries
    # For each key, check:
    #   1. Is it required but missing from config?
    #   2. Is it present but has the wrong type?

    return errors


# ─── Tests ───────────────────────────────────────────────────────────────────

def test_valid_config():
    schema = {
        "host":    {"type": str,   "required": True},
        "port":    {"type": int,   "required": True},
        "debug":   {"type": bool,  "required": False},
        "timeout": {"type": float, "required": False},
    }
    config = {"host": "localhost", "port": 8080, "debug": False, "timeout": 30.0}
    assert validate_config(config, schema) == [], "Expected no errors for valid config"

def test_missing_required():
    schema = {
        "host": {"type": str, "required": True},
        "port": {"type": int, "required": True},
    }
    config = {"host": "localhost"}
    errors = validate_config(config, schema)
    assert "Missing required key: 'port'" in errors

def test_missing_optional_ok():
    schema = {
        "host":  {"type": str,  "required": True},
        "debug": {"type": bool, "required": False},
    }
    config = {"host": "example.com"}
    errors = validate_config(config, schema)
    assert errors == [], "Optional missing keys should not cause errors"

def test_wrong_type():
    schema = {"port": {"type": int, "required": True}}
    config = {"port": "8080"}  # string instead of int
    errors = validate_config(config, schema)
    assert "Key 'port': expected int, got str" in errors

def test_bool_not_valid_int():
    """bool is a subclass of int in Python — we treat it as a type error."""
    schema = {"port": {"type": int, "required": True}}
    config = {"port": True}
    errors = validate_config(config, schema)
    assert "Key 'port': expected int, got bool" in errors, \
        "True/False should not be accepted as int values"

def test_multiple_errors():
    schema = {
        "host":  {"type": str,  "required": True},
        "port":  {"type": int,  "required": True},
        "debug": {"type": bool, "required": False},
    }
    config = {"host": 1234, "debug": "yes"}  # host wrong type, port missing, debug wrong type
    errors = validate_config(config, schema)
    assert len(errors) == 3
    assert "Key 'host': expected str, got int" in errors
    assert "Missing required key: 'port'" in errors
    assert "Key 'debug': expected bool, got str" in errors

def test_extra_keys_ignored():
    """Keys in config that aren't in schema should be silently ignored."""
    schema = {"host": {"type": str, "required": True}}
    config = {"host": "localhost", "extra_setting": 42}
    errors = validate_config(config, schema)
    assert errors == [], "Extra keys should be ignored"


if __name__ == "__main__":
    tests = [
        test_valid_config,
        test_missing_required,
        test_missing_optional_ok,
        test_wrong_type,
        test_bool_not_valid_int,
        test_multiple_errors,
        test_extra_keys_ignored,
    ]
    passed = 0
    for t in tests:
        try:
            t()
            print(f"  ✓ {t.__name__}")
            passed += 1
        except AssertionError as e:
            print(f"  ✗ {t.__name__}: {e}")
        except Exception as e:
            print(f"  ✗ {t.__name__}: unexpected error — {e}")

    print(f"\n{passed}/{len(tests)} tests passed")

File: challenges/2026-03-03-json-config-validator.ts
/**
 * 🧩 Daily Challenge: 2026-03-03
 * Title: JSON Config Validator
 * Language: TypeScript
 * Difficulty: Beginner
 *
 * ## Problem
 *
 * You're building a tool that reads JSON configuration files. Before using the
 * config, you need to validate that it has the required fields with the correct
 * types.
 *
 * Write a function `validateConfig` that takes an unknown parsed JSON value and
 * checks whether it matches a given schema. The schema is a simple object
 * describing required keys and their expected types.
 *
 * Supported type strings: "string", "number", "boolean", "object", "array"
 *
 * The function should return an object with:
 *   - `valid`: boolean — whether the config passes all checks
 *   - `errors`: string[] — a list of human-readable error messages for each violation
 *
 * ## Examples
 *
 * ```ts
 * const schema: Schema = {
 *   host: "string",
 *   port: "number",
 *   debug: "boolean",
 * };
 *
 * validateConfig({ host: "localhost", port: 8080, debug: false }, schema);
 * // => { valid: true, errors: [] }
 *
 * validateConfig({ host: "localhost", port: "8080" }, schema);
 * // => {
 * //   valid: false,
 * //   errors: [
 * //     'Field "port" expected type "number" but got "string"',
 * //     'Missing required field "debug"',
 * //   ]
 * // }
 *
 * validateConfig("not an object", schema);
 * // => { valid: false, errors: ["Config must be a non-null object"] }
 *
 * const nestedSchema: Schema = { tags: "array", metadata: "object" };
 * validateConfig({ tags: ["a", "b"], metadata: { version: 1 } }, nestedSchema);
 * // => { valid: true, errors: [] }
 * ```
 *
 * ## Starter Template
 */

type TypeName = "string" | "number" | "boolean" | "object" | "array";
type Schema = Record<string, TypeName>;

interface ValidationResult {
  valid: boolean;
  errors: string[];
}

function validateConfig(config: unknown, schema: Schema): ValidationResult {
  // TODO: Implement this function
  // Step 1: Check that config is a non-null object (and not an array)
  // Step 2: For each key in the schema, check:
  //   - Does the key exist in config?
  //   - Does the value match the expected type?
  //   - Special case: "array" should use Array.isArray(), "object" should
  //     exclude arrays and null
  // Step 3: Collect all errors and return the result
  throw new Error("Not implemented");
}

// ============================================================
// HINTS (scroll down)
//
//
//
//
//
//
//
//
//
//
//
//
// Hint 1: Use `typeof` for basic type checks. But remember:
//   typeof null === "object" and typeof [] === "object"
//   So you need special handling for "array" and "object".
//
// Hint 2: To check if a key exists on an unknown object, first
//   narrow the type to Record<string, unknown> then use `in` or
//   hasOwnProperty.
//
// Hint 3: A helper function like getTypeName(value: unknown): TypeName
//   can keep your code clean. It returns "array" for arrays,
//   "object" for plain objects, and typeof for the rest.
// ============================================================

// --- Tests (run with: npx ts-node 2026-03-03-json-config-validator.ts) ---

function runTests() {
  const schema: Schema = { host: "string", port: "number", debug: "boolean" };

  // Test 1: valid config
  const r1 = validateConfig({ host: "localhost", port: 8080, debug: false }, schema);
  console.assert(r1.valid === true, "Test 1 failed: should be valid");
  console.assert(r1.errors.length === 0, "Test 1 failed: should have no errors");

  // Test 2: wrong type + missing field
  const r2 = validateConfig({ host: "localhost", port: "8080" }, schema);
  console.assert(r2.valid === false, "Test 2 failed: should be invalid");
  console.assert(r2.errors.length === 2, `Test 2 failed: expected 2 errors, got ${r2.errors.length}`);

  // Test 3: not an object
  const r3 = validateConfig("not an object", schema);
  console.assert(r3.valid === false, "Test 3 failed: should be invalid");
  console.assert(r3.errors.length === 1, "Test 3 failed: should have 1 error");

  // Test 4: null input
  const r4 = validateConfig(null, schema);
  console.assert(r4.valid === false, "Test 4 failed: null should be invalid");

  // Test 5: array and object types
  const r5 = validateConfig(
    { tags: ["a", "b"], metadata: { v: 1 } },
    { tags: "array", metadata: "object" }
  );
  console.assert(r5.valid === true, "Test 5 failed: should be valid");

  // Test 6: array should not match "object" and vice versa
  const r6 = validateConfig(
    { tags: { a: 1 }, metadata: [1, 2] },
    { tags: "array", metadata: "object" }
  );
  console.assert(r6.valid === false, "Test 6 failed: swapped types should fail");

  // Test 7: extra fields are fine (we only check required ones)
  const r7 = validateConfig({ host: "x", port: 1, debug: true, extra: "ok" }, schema);
  console.assert(r7.valid === true, "Test 7 failed: extra fields should be allowed");

  console.log("All tests passed (or check assertions above)!");
}

runTests();
